presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusSplitManager.java

```diff
@@ -0,0 +1,156 @@
+        if (table == null) {
+            throw new TableNotFoundException(tableHandle.toSchemaTableName());
+        }
+        List<ConnectorSplit> splits = new ArrayList<>();
+        generateTimesForSplits(Instant.now().toEpochMilli()).forEach(time -> {
+            try {
+                splits.add(new PrometheusSplit(buildQuery(prometheusClient.config.getPrometheusURI(), time, table.getName())));
+            }
+            catch (URISyntaxException e) {
+                throw new UnsupportedOperationException(e.getMessage());
+            }
+        });
+        return new FixedSplitSource(splits);
+    }
+
+    // URIBuilder handles URI encode
+    private URI buildQuery(URI baseURI, String time, String tableName)
+            throws URISyntaxException
+    {
+        List<NameValuePair> nameValuePairs = new ArrayList<>(2);
+        nameValuePairs.add(new BasicNameValuePair("query",
+                tableName + "[" + prometheusClient.config.getQueryChunkSizeDuration() + "]"));
+        nameValuePairs.add(new BasicNameValuePair("time",
+                time));
+        return new URIBuilder(baseURI.toString())
+                .setPath("api/v1/query")
+                .setParameters(nameValuePairs).build();
+    }
+
+    /**
+     * Get the start times in seconds that divide up the query using the config value:
+     * maxQueryRangeDuration
+     * queryChunkSizeDuration
+     * The times will be used in queries to Prometheus like: `http://localhost:9090/api/v1/query?query=up[3w]&time=1568229904"`
+     * ** NOTE: Prometheus instant query wants the end time specified, you get duration before that time returned
+     *
+     * @return list of start times as epoch seconds, like ["1568053244", "1568926595"]
+     */
+    private List<String> generateTimesForSplits(long nowTimeInMillis)
+    {
+        long maxQueryRangeDuration = timeUnitsToSeconds(prometheusClient.config.getMaxQueryRangeDuration());
+        long splitChunkSizeDuration = timeUnitsToSeconds(prometheusClient.config.getQueryChunkSizeDuration());
+        long nowSeconds = TimeUnit.MILLISECONDS.toSeconds(nowTimeInMillis);
+        int numChunks = (int) Math.ceil((double) maxQueryRangeDuration / (double) splitChunkSizeDuration);
+        long firstEndTime = nowSeconds - maxQueryRangeDuration + splitChunkSizeDuration;
+
+        return IntStream.range(0, numChunks).mapToObj(n -> {
+            long endTime = firstEndTime + n * splitChunkSizeDuration;
+            if (endTime > nowSeconds) {
+                endTime = nowSeconds;
+            }
+            return endTime;
+        }).map(String::valueOf).collect(Collectors.toList());
+    }
+
+    /**
+     * s - seconds
+     * m - minutes
+     * h - hours
+     * d - days
+     * w - weeks
+     * y - years
+     *
+     * @param timeWithUnits
+     * @return long for the seconds represented by the passed timeWithUnits
+     */
+    static long timeUnitsToSeconds(String timeWithUnits)
```

Comment:
I can't help but feel like a java library exists for this. I'm somewhat concerned about dealing with time like this as time can be a bit wonky and also, parsing like this can have edge cases. Looking around I see https://stackoverflow.com/questions/6403851/parsing-time-strings-like-1h-30min using joda-time which Presto already uses. Additionally, I believe Presto already has some configuration options in terms of durations like `1d` and `30s`, etc, so I bet you can find what those use too.

Reply:
Yah airlift's Duration does it. Done in: 0c4dbd3